File: checkout-ruby/lib/checkout_item.rb
class CheckoutItem

  attr_reader :item, :rule, :count

  def initialize(item, rule)
    @item = item
    @rule = rule
    @count = 0
  end

  def add_item
    @count +=1
  end

  def get_price
    if !rule.nil? && rule.is_eligible_for_discount?(count)
      rule.value_type.get_discounted_price(item.price) * count
    else
      item.price * count
    end
  end

end

File: checkout-java/src/main/java/com/vinterior/checkout/models/CheckoutItem.java
package com.vinterior.checkout.models;

import lombok.Getter;

@Getter
public class CheckoutItem {

    private int count = 0;

    private Item item;

    private PromotionalRule discountRule;

    public CheckoutItem(Item item, PromotionalRule rule) {
        this.item = item;
        this.discountRule = rule;
    }

    public void addItem() {
        count++;
    }

    public double getPrice() {
        double price = 0.0;
        if(null != discountRule && discountRule.isEligibleForDiscount(count)) {
            price = discountRule.getValueType().getDiscountedPrice(item.getPrice()) * count;
        } else {
            price = item.getPrice() * count;
        }
        return price;
    }


}
